### ISSUE TITLE
Feature request: allowing users to manually select Region / Video Mode (PAL, NTSC, AUTO)

### ISSUE THREAD

**Issue by Ryunam** [libretro/beetle-psx-libretro#67]
This is to kindly inquire about the possibility of exposing an option to select the Region / Video Mode in the Core Options of beetle-psx-libretro, thus allowing the user to either force a specific format (PAL / 50hz or NTSC / 60hz) or leave it up to the automatic region detection normally performed by the emulator.

The reason behind this request lies in being able to circumvent the technical limitations that affected the PAL-exclusive localizations of several classics.
Back in the day, PAL PS1 games were released on the market in a crippled form that capped them at 50hz, making the overall framerate 17,5% slower if compared to their respective NTSC counterparts. Furthermore, because of the intrinsic difference between the two formats and the different number of lines displayed, PAL versions were often plagued with black borders.

While it appears that the latter effect is mitigated in emulation, the former is still firmly in place. 
That is because, as far as I understand, Mednafen PSX faithfully reads the regional code of each game and then proceeds to run it with the corresponding Region / Video Mode, but this makes it so that some German, French, Italian and Spanish translations are inevitably hindered by the slower framerate. 
Being able to choose between modes would allow a wider array of options, while keeping the goal of accuracy.

Thanks in advance!

**Comment by simias** [comment]
That sounds trivial enough to implement, a "force PAL/NTSC" option in the core. I'll give it a try.

**Comment by simias** [comment]
On 2nd thought just changing the detected video format won't be enough, you'd also have to hijack the game to prevent it from setting the PAL bit in the GPU DisplayMode (all PSX GPUs can output both PAL and NTSC, there's a register to configure this behaviour).

**Comment by Ryunam** [comment]
I see. In fact one thing I want to highlight here is that I had attempted to patch my PAL localized games to NTSC through some tools available around (which I assume have the effect of changing the GPU DisplayMode you referenced), but since Mednafen automatically detects the video format in an automatic way they would play in their usual crippled speed (50hz).

I wonder if in the meantime the option you have mentioned in your first post could be implemented anyway, so as to test how it behaves with NTSC-patched images. 

Or, another idea that comes to my mind, would it be possible to make it so that when you activate a forced Region / Video mode from the Core options, it would also impose the same configuration for the GPU bit?

**Comment by needo2** [comment]
> PAL versions were often plagued with black borders

oddly enough, Spyro 1,2,3 PAL games cover whole screen while NTSC are padded. I'd play them in PAL if framerate wasn't crippling

**Comment by simias** [comment]
Spyro PAL runs at 30fps, same framerate as the NTSC version:

![retroarch-0621-144157](https://cloud.githubusercontent.com/assets/7925386/16229779/90721b18-37be-11e6-9129-9a683d0d128b.png)

It just has a slightly higher vertical resolution.

I actually had trouble finding PAL ports that ran slower that the NTSC version, it seems I've been lucky with the few games from my collection I've ripped. I should try a final fantasy game, they're quite notorious for their crappy PAL ports.

**Comment by needo2** [comment]
Those green bars on bottom and right appear at NTSC version for me.

I put NTSC vs PAL side by side RetroArch Mednafen PSX (Software)

                | PAL                        | NTSC           
-------------|--------------------------|--------
**Display**      | 4:3                         | 4:3 letterbox 
**Framerate** | 29.97 feels like 20 | 29.97            
**Artifacts**     | none                      | Bars on bottom and right depending on the screen you are

**Comment by simias** [comment]
Oh you're right, the internal FPS counter is broken currently, it always assumes NTSC. Let me try to fix that.

Spyro PAL really runs at 25fps.

**Comment by rz5** [comment]
@simias, is '25 fps' something that Spyro (PAL) requests and something we can intercept? From my experience messing with the rustation renderer I know there this:
https://github.com/libretro/beetle-psx-libretro/blob/master/rustation-libretro/src/retrogl/retrogl.cpp#L327

I guess this would be one of the places where we'd read a hypothetical Region/Video mode core option and force some mode if it's not on 'auto'.

**Comment by simias** [comment]
@rz5 I think you'd have to do it earlier than that. You can't completely pretend that the game is NTSC however (because things like BIOS detection will fail otherwise).

However you have to lie to the GPU by passing "is_pal = false" to the GPU constructor at least, and you also have to tell the GPU to ignore the game's attempt at setting the PAL video format.

However you'll still want to display the full PAL range if you don't want to crop things out. Not sure how that'll go. I'll give it a try I guess.

**Comment by simias** [comment]
So the problem is that the internal renderer poops itself if I lie about the game rendering in PAL (it's not really surprising, it messes up its internal counters).

The GL renderer however works at a higher level (I don't actually count lines and whatnot) so it works somewhat better, unfortunately it seems that the software renderer commits suicide and stops uploading the textures so I get an untextured Spyro, albeit running at 30fps:

![spyro-pal-60](https://cloud.githubusercontent.com/assets/7925386/16249599/62097fa6-3817-11e6-8e05-799cd0dad77e.png)

I don't really get why the software renderer refuses to proceed. At any rate I don't see how we could get the software renderer to render PAL at 60fps without cropping some lines since it wouldn't fit the timings. Maybe by fudging the timings somewhat but I don't feel comfortable doing that, it's quite the state machine we have there.

**Comment by needo2** [comment]
What about running NTSC with PAL resolution? Can the letterbox get removed?

**Comment by simias** [comment]
The black bars are here because of mednafen's emulation of a real monitor, this way it can implement things like the screen shaking in some games: https://www.youtube.com/watch?v=a-ovOoISyLA

The GL renderer doesn't implement that currently so you don't have black bars, all you need to do is had something similar to the widescreen hack to increase the vertical angle to match the PAL resolution. It would probably break with some games though. That wouldn't increase the actual resolution but the camera would match and you could just run the game at 2x or more to effectively increase the resolution to compensate.

I'm not sure it's worth bothering with it though, why not just play the NTSC version with increased internal res? The camera will be a bit "flatter" than on PAL counterparts (for proper ports at least) but that's the only downside really.

**Comment by Ryunam** [comment]
First of all, I'd like to thank simias for their patience and dedication. It is nice to hear that the Internal FPS Counter was fixed: when I first read your previous reports stating that PAL games had the same framerate if compared to their NTSC variants, I was a bit perplexed! Unfortunately this is an issue that European players have been living with for ages, long before Video Mode selectors were implemented in more modern systems (ie from the PS2 onwards), so the difference in smoothness and actual game speed is very noticeable while playing.

With regards to the Video / Region Code I've been thinking a few things:

1) I had managed to find a way to play my PAL Italian version of Final Fantasy VIII at perfectly smooth NTSC speeds (60hz) on the emulator ePSXe. I basically activated an option that would force the emulator to always use an NTSC BIOS (either US or JPN) and tell the game that it was running on a non-PAL console. Then, on the video plugin side, there was an option to indicate '60fps' as the minimum framerate. 
As far as I remember this combination of settings worked well: the game played exactly like my Japanese copy in terms of speed / smoothness and audio and video appeared to be well synchronized.
I know mednafen doesn't use plugins and is much more accurate, but I wonder whether this method could be useful to devise a workaround.

2) The way I usually force NTSC on the actual hardware (a modded PS1 PAL system) is by using one of the many Boot disks that float around the internet. The most effective one is ImportPlayer / ImportPlayer Light: you start the system with this disk inside the tray, you then select the actual Video Mode that you want to force, then you swap the disks and put the game you want to play with the forced Region / Video Mode. 
Last time I tried using these Boot Disks on mednafen (by using all the proper commands to cycle the tray and swap the respective images) it would simply hang, as referenced in this thread I made a few months ago on the libretro boards: [http://libretro.com/forums/showthread.php?t=4913](url)
I wonder if the commands used by these Disks could be useful to implement a similar option at a core-wide level.

3) There are also many patches available for PAL PS1 titles that add a _Video Mode selector_ right before starting the actual game. They work very well on actual hardware (albeit once again modded).
The problem with these modifications is that on Mednafen, when forcing NTSC through these patches, the game displays a black screen even though you can still hear the sound going. 
This is also referenced in another issue that was opened here on github: [https://github.com/libretro/beetle-psx-libretro/issues/3](url). 
RagnaStrife says he managed to play these modded games with NTSC speeds once he forced the NTSC video mode on standalone Mednafen.

Hope this is helpful one way or another!

**Comment by needo2** [comment]
If it knows the display resolution of NTSC, perhaps a zoom feature might do it, scale the screen preserving aspect ratio?

**Comment by simias** [comment]
@Ryunam for your methods 2 and 3 I assume that those methods completely switch to NTSC video mode, i.e. you also lose a bit of vertical resolution?

The difficulty is to run at 60fps while keeping the PAL resolution I think (PAL60).

**Comment by Ryunam** [comment]
@simias : Yes, I can confirm that. My CRT TV switches to NTSC completely with both methods 2 and 3. Framerate improves drastically since at that point the game is running at NTSC speeds (60hz), but the vertical resolution is indeed reduced.

Still, it would be great to actually have this option on Mednafen, for all the reasons already indicated.

**Comment by simias** [comment]
I think I have a working patch for that:

![spyro](https://cloud.githubusercontent.com/assets/7925386/16353086/84f5f444-3a75-11e6-84e8-805b56b202c4.png)

It requires a lot more testing though, I wish we had an "experimental" core for that stuff.

Anyway, I'm going to turn it into a core option and you'll tell me if it does what you want.

**Comment by simias** [comment]
Just pushed the option, it seems to mess up the SCEE boot logo for some reason (but it's already glitchy with beetle currently for some reason). Spyro, MSG and Rollcage all seem to work at 30FPS when forcing the NTSC videomode with no obvious glitch. That being said I'm bad at noticing framerate stuff so let me know if it doesn't work properly.

**Comment by simias** [comment]
Oh and you need to restart the emulator for changes to this option to take effect, it's not applied directly.

**Comment by needo2** [comment]
Spyro NTSC on PAL resolution, x2 internal, on software rendering runs very slow without Threaded Video and with it, there are slow downs from time to time. And opengl is always very slow for me regardless of the ntsc-pal


Cropped a bit much (doesn't bother me)
![retroarch-0625-095053](https://cloud.githubusercontent.com/assets/9450683/16355284/fefd0530-3aba-11e6-8aae-4c7816aefae6.png)

**Comment by Ryunam** [comment]
Thanks for the patch, @simias! Unfortunately there seem to be some issues with the current implementation. 
I have performed a few tests so far, namely trying to force my PAL Italian copy of Final Fantasy VIII to NTSC and the results I've got are quite odd.

**With the software renderer**

- The game boots, but does not display anything whatsoever, not even the SCEE logo. Only audio is heard.
- Internal FPS shows as _59.97_, but external FPS of Retroarch shows as _50_.

**With the opengl renderer**

- The game boots, video and audio are shown, although the renderer slows down and stutters frequently even at 1x default parameters, to the point of making it unplayable (might be entirely unrelated to this Video Mode fix, as @needo2 pointed out).
- Once again, internal FPS shows as _59.97_, but external FPS of Retroarch shows as _50_.
- If I press F to alternate between Exclusive Fullscreen and Windowed mode, the video disappears completely and the game behaves exactly as if I used the software renderer (only audio is heard). 
- After video blacks out, if I go back to the XMB Menu, Retroarch's text appears garbled and full of artifacts.
- If I apply a shader after emulation has started, Retroarch crashes.

**Comment by simias** [comment]
Mmmh. That's odd. I did only try the GL shader, I'll try with software mode. How do you measure the "external FPS" exactly? Using RA's builtin "Display Framerate" option I was running at 60fps in Spyro.

**Comment by saftle** [comment]
@simias I run RetroArch through Steam with the in-game overlay FPS enabled. Or others use MSI Afterburner, or Fraps.

**Comment by Ryunam** [comment]
Yes, sorry for the imprecise wording, @simias . Where I said  'external FPS' in my previous post, I referred to the value reported by Retroarch's builtin Display Framerate.

I have retried once again with the OpenGL renderer, going past the initial starting Menus and once again I can confirm my initial results:
- Internal FPS: 59.97 in menus, 29.97 in-game (as it should be in NTSC)
- Builtin Display Framerate: 50fps
- Alternating between Fullscreen and Windowed mode makes the video disappear, just like in software mode.

**Comment by simias** [comment]
so you mean that RA reports running at 60fps but it really runs at 50?

**Comment by Ryunam** [comment]
Basically, with PAL Final Fantasy VIII forced to NTSC, Mednafen's internal FPS is reporting 59.97 / 29.97 (NTSC), but RA's display framerate displays 50fps. This happens in both software and OpenGL renderers, so it's still rendering in PAL / 50hz.

The only difference between both renderers is that:
- with software rendering video does not work anymore when forcing NTSC. Only audio
- with OpenGL both video and audio work, but the other issues still happen and the game still renders in PAL / 50fps

**Comment by simias** [comment]
Okay, I'll give PAL FFVIII a try then. I tried with Rollcage, MGS and Spyro.

**Comment by Ryunam** [comment]
I got a few screen grabs to further demonstrate the issues with Final Fantasy VIII (PAL):

**With the Software renderer**

Internal FPS: 
![retroarch-0625-204149](https://cloud.githubusercontent.com/assets/17614150/16358587/77358bec-3b17-11e6-95be-6ad2ee1e9ecc.png)

Retroarch's Display FPS:
![retroarch-0625-204315](https://cloud.githubusercontent.com/assets/17614150/16358588/8a0f9ec4-3b17-11e6-824d-e0e73bbada0b.png)

**With the OpenGL renderer**

Internal FPS:
![retroarch-0625-205333](https://cloud.githubusercontent.com/assets/17614150/16358605/32facd2e-3b18-11e6-935a-d2ae36e370a8.png)

Retroarch's Display FPS:
![retroarch-0625-205253](https://cloud.githubusercontent.com/assets/17614150/16358608/384f9cfa-3b18-11e6-86d9-3d7a41743dd6.png)

**With the OpenGL renderer, after alternating between Fullscreen and Windowed**

![retroarch-0625-205359](https://cloud.githubusercontent.com/assets/17614150/16358610/545f3144-3b18-11e6-95c5-a1556ede4e63.png)

**Comment by Ryunam** [comment]
Did you find out anything new about this, @simias? I've seen that the previous implementation was reverted as it seemed to break PAL emulation completely.

**Comment by simias** [comment]
No, I'm not sure why it even broke anything to begin with. I'll have to give it an other try later.

**Comment by E38IK** [comment]
Is there any interest in resuming work on a feature like this?

There have been a raft of improvements to beetle since this was opened, and it's by far the best PS1 emulator around now.
Not being able to run PAL games at 60hz is one of the few remaining blemishes. I understand this is a problem with mednafen-psx at its core, but maybe something could be done about it from this side.
There are a number of games where, content-wise, the PAL version is the better one, but the associated slowdown (and judder on non-VRR displays or displays with broken 50hz modes) is a dealbreaker.

**Comment by hizzlekizzle** [comment]
Is it any more complicated than adding a core option for PAL60 and changing:
`float fps = is_pal ? FPS_PAL : FPS_NTSC;`
to
`float fps = (is_pal && !PAL60) ? FPS_PAL : FPS_NTSC;`

**Comment by E38IK** [comment]
Presumably. Currently, mednafen-psx (standalone as well as beetle-psx) will autodetect the region of a game and forcibly ignore any PAL2NTSC patches you may have applied to the images. Those patches do work on actual hardware as well as various other emulators (like pSX).
Mednafen-psx exposes an option to disable region auto detection (which isn't accessible via beetle-psx as far as I can tell); however, it will then throw a video mode mismatch error and not display any video output. As this runs counter to the way actual hardware behaves, I would assume there is a reason for this and something else would break.

**Comment by bigboo3000** [comment]
I also tried hard-patching .bin image files for pal60 with no luck with beetle-psx, now I understand why. A core option to force PAL games to PAL60 would be so great.

Maybe something related to this?
https://github.com/libretro/beetle-psx-libretro/issues/376

**Comment by E38IK** [comment]
> Maybe something related to this?
> #376

While that technically gets you 60hz, this will really mess with the audio. It's unfortunately not a very useful solution.

**Comment by BardiBard** [comment]
Is there no way to make this work on Beetle PSX HW?
If you have a patched PAL2NTSC game, that the core recognozes it and just treats it as an NTSC game?
On the ReARMed core it works and the patched PAL game will run as NTSC no problem, as long as you manually set the region in the options tab to "NTSC".
But that core doesn't have those great gaphics enhancement options as Beetle PSX HW does.

In the Discord I've read that this is some kind of issue with upstream mednafen.
It uses the disc's serial to determine if it is NTSC or PAL, regardless on if it has been patched with Zapper2K/PAL4U.

Maybe if there was a manual option to set PAL / NTSC / AUTO on region detection, it would work with patched PAL games at least.

Forcing unpatched games into NTSC seems to be the main problem here as far as I can tell.
But maybe the result will be better by just trying how it'll work with patched PAl2NTSC games.

This .diff file here has been created by the Discord user radius from the libretro Discord Server:
[region.zip](https://github.com/libretro/beetle-psx-libretro/files/2692052/region.zip)
Maybe it'll help in this effort.
He said that if you are able to build a core yourself, this might do the trick with patched games.

Whoever is still working on this, maybe give it a try.
It would help with a lot of problems PAL games come with such as de-synched 3D ingame cutscenes or that "teleport"-effect on video scenes, where it jumps ahead 5 frames every second, if the core would allow patched games to work as NTSC.

**Comment by Papermanzero** [comment]
I only want to emphasize that the realisation of the features needs to differentiate between two settings. The color component (PAL, NTSC, SECAM) and the Refresh Rate (25Hz, 30Hz, 50Hz, 60Hz)
The details can be read in the linked (and closed) issue 
https://github.com/libretro/beetle-psx-libretro/issues/503

**Comment by rz5** [comment]
@Papermanzero - You linked to a forum thread discussing ways to force PS1/PS2 hardware into different modes.

So an interesting thing to try out is to attempt the same thing but with beetle-psx.
Put one of those boot discs in beetle-psx's virtual disc tray (just launch the .cue/.chd file), do the mode diddling stuff and when the mode switching program asks you to insert a game CD, replace the virtual mode switching disc with a game disc (Quick Menu > Disk Control > Disk Image Append > game.chd; Disk Control > Disk Cycle Tray Status).

If that works, great. It's a clue on how to get an implementation going. It if does not work, that's great too. We can go to upstream's bug tracker and file that as a bug.

**Comment by trostboot** [comment]
That doesn't work. Like it was alluded to by E38IK a few posts up in the issue, mednafen/beetle will simply ignore these kinds of patches - whether they're on-the-fly or applied to the image. Just to be sure, I checked it out using ImportPlayer Light, and it doesn't work.

**Comment by trostboot** [comment]
Actually, I may be mostly wrong. Using beetle and a German Final Fantasy IX image with a US region ImportPlayer Light it does appear to run correctly at 60hz - using the EU IPS version does not, it will stay at 50hz.
For this to work on upstream mednafen, regio auto detect has to be on.

**Comment by darkcloudl** [comment]
I actually use the method explained by rz5 since RA 1.7.5 in beetle psx hw(vulkan) with 'skip bios' off and works, i use "ImportPlayer Light v2.1 [EURO]" and force to ntsc the pal games exhumed(eur version of powerslave) RRT4, everybody's golf(eur version of hot shots golf 1),dino crisis 1(in this you even can adjust ingame the image with the L1+R1+L2+R2+Up/down keys; keep in mind this game requires sbi fie because LibCrypt protection if someone else want to test) and others; didn't work with metal gear solid(SLES-01734, stuck in anti piracy message screen). The games displays a 60hz image and even you can notice the difference (especially with RRT4 and Exhumed) but the 'display internal FPS' option in the core still displays pal framerate (the RA 'display framerate' option and even msi afterburner shows 60 fps).

**Comment by Papermanzero** [comment]
Interesting that the boot discs are correctly emulated.

**Comment by BardBun** [comment]
Is there any progress being made on this?

To me this is a very important issue, so many PAL games usually are the definitive version of a game but are crippled with 50 FPS, causing FMV's to always skip frames.

**Comment by Papermanzero** [comment]
Also agree. Several PAL versions are locked on 50Hz but contain various bonuses like several languages and content. It would be really nice to speed them up to their original speed.

**Comment by simias** [comment]
> I only want to emphasize that the realisation of the features needs to differentiate between two settings. The color component (PAL, NTSC, SECAM) and the Refresh Rate (25Hz, 30Hz, 50Hz, 60Hz)

I'm not sure I understand what you mean by "color component", the actual color encoding isn't relevant for the emulator because it doesn't actually simulate the PAL/NTSC/SECAM signal, it stops at the RGB buffer being sent to the output. Therefore as far as the emulator is concerned PAL and SECAM are the exact same thing and the only difference between PAL and NTSC are different timings.

**Comment by Papermanzero** [comment]
I only wanted to say that PAL, NTSC and Secam are color coding information. They are independent of the refresh rate. Means, you can have pal in 50hz and 60hz. in theory you could also limit ntsc to 50hz but this would be outside of the standard.

**Comment by simias** [comment]
That's true for PAL and Secam (same video timings, different color coding) but NTSC does have different timings. PAL 60 exists but that's effectively a different standard that came after the fact. The original PSX doesn't support PAL 60.

It's not just nitpicking, as far as the PSX is concerned there's a very significant timing difference between PAL and NTSC console. Even the hardware GPU clock is different!

That being said forcing a different standard shouldn't normally be very difficult, the only "problem" with mednafen/beetle is that it emulates some low-level details about scanlines and blankings that would probably cause some issues with a naive implementation. Since mednafen outputs one line at a time if you increase the framerate without reducing the line count you'll have the problem since there won't be enough time to draw everything correctly.

Intuitively the solution would be to change Mednafen's rendering to allow rendering one frame at a time instead of one line at a time. This way the number of lines being rendered won't directly be tied to the timings.

I'm currently in the process of writing my own renderer based on mednafen's so I'll try to keep that in mind. Slow and steady.

**Comment by simias** [comment]
I made a new attempt in my branch to see if I could cheat by telling the GPU code to use "shorter' lines (timing-wise, not resolution-wise) for PAL: https://github.com/simias/beetle-psx-libretro/tree/dev/pal-60fps

It does seem to desync the audio from time to time in Spyro, mainly during cinematics. If somebody wants to give it a try (there's no option, it's just hardcoded for every PAL game).

**Comment by simias** [comment]
Yet an other possibility would be to speed up the entire system by a factor of 30/25 to fake the 30fps, but have the SPU run at the normal speed so that the sound wouldn't be distorted.

**Comment by Ryunam** [comment]
Thanks @simias, I'm glad that this issue is still getting some attention after 4 years! I can only echo once again the other commenters: this feature would be extremely handy since so many interesting improvements and translations are locked behind the slowdown associated with the PAL versions.

I'll test your branch in a second, though just as a comparison I will note that, if you use a boot disc with a PAL / NTSC force boot feature such as Import Player Light, PAL games will typically run at the NTSC framerate without any desynchronization (both on the Beetle PSX core - based on the few tests I've run before - and on real hardware).

The only issue with that method is that some content will end up displaying some anomalies in the geometry of the picture for some reason (even on real hardware): for instance, when booting the European version of FFVIII in NTSC mode with the Import Player Light disc, anytime you open the menu the whole picture will be shifted upwards by a significant amount.

**Comment by Ryunam** [comment]
Okay @simias, I tested your build from the dev/pal-60fps branch, but it doesn't seem to show any difference compared to the normal core.
All PAL content I have tried plays at the usual PAL speed (50Hz) and with black borders, which is another typical limitation of PAL PS1 games.

Interestingly, on the other hand if you use the PCSX-ReArmed core and set the Region to NTSC in the core options, that core will play any and all PAL content at 60hz and without black borders at the top and bottom. I understand though that the two cores have an entirely different codebase, so I don't know if it's a useful comparison.

The games I've tried are:
- Final Fantasy VIII
- UEFA Euro 2000
- Bishi Bashi Special

**Comment by hizzlekizzle** [comment]
mednafen has a region-picking (psx.region_default) option, but I'm assuming it doesn't do what we need to do? I've messed around with that option in beetle-psx but haven't gotten it to actually *do* anything.

**Comment by simias** [comment]
> Okay @simias, I tested your build from the dev/pal-60fps branch, but it doesn't seem to show any difference compared to the normal core.

So that's because I was attempting the homeopathic approach, given that I just noticed that I created the branch but forgot to push the actual commit making the change. So I guess you passed the double blind test and placebo emulation doesn't actually work.

> Interestingly, on the other hand if you use the PCSX-ReArmed core and set the Region to NTSC in the core options, that core will play any and all PAL content at 60hz and without black borders at the top and bottom.

Well if somebody can describe precisely what PCSX-ReArmed is doing I might be able to port it into beetle.

From that description it means that the timings shift completely to NTSC and you end up with true PAL60? I was explicitly trying not to do that because it would break with games that output actual PAL without bars (those would get cropped) but that's definitely doable too.

**Comment by simias** [comment]
I actually pushed the commit now by the way, if you feel like testing again. You should definitely still get black bars but the framerate should go from 50 to 60FPS for PAL games (and the timings for NTSC should be subtly messed up, so don't worry if that doesn't work correctly).

**Comment by Ryunam** [comment]
> So that's because I was attempting the homeopathic approach, given that I just noticed that I created the branch but forgot to push the actual commit making the change. So I guess you passed the double blind test and placebo emulation doesn't actually work.

Haha, that's good, I have a long history of chasing NTSC imports back in the day because they lacked both the slowdown and borders that plagued PAL releases, so I guess I'm the right person to test this. :)

> From that description it means that the timings shift completely to NTSC and you end up with true PAL60? I was explicitly trying not to do that because it would break with games that output actual PAL without bars (those would get cropped) but that's definitely doable too.

Yes, what ReArmed is doing seems to be exactly that: upon setting the "Region" to NTSC, it is forcing PAL games to behave as if they were NTSC. The black bars disappear and the framerate is adjusted accordingly.

One way to accommodate all use cases would be to have a core option for Beetle ("Force NTSC timings") with the following values selectable:

- OFF (default, works the way current code behaves and does not force anything);
- Framerate only (what your commit is doing - increasing the framerate only, without messing with the black bars so that properly-converted PAL titles can avoid being cropped);
- Force everything (so that both the framerate and the black bars are forced to NTSC. This would be the way PCSX-ReArmed is handling it).

**Comment by Ryunam** [comment]
@simias, I tested your commit and I am happy to confirm that it is working as intended. PAL games now play at 60fps.

One thing I noticed is that some games (like Final Fantasy VIII) do not exhibit any desynchronization at all. Both normal gameplay and the FMV cinematics play perfectly fine and never lose sync.
Other titles (for example Bishi Bashi Special among the games I've tested) seem to have the issues you experienced with Spyro. The audio is cut off at certain points because the video is now faster and the sound doesn't keep up.

Given that this is most likely caused by the different methods used when the developers converted each title from NTSC to PAL, perhaps we would need ideally another core option that lets the user adjust the SPU speed? I am not sure about this, the only evidence I can bring to the table is that some content does remain in sync, some other content has issues with this (and this probably happens on real hardware too, with the import bootdiscs).

**Comment by simias** [comment]
Thank you for taking the time to test it.

Adjusting the SPU speed would result in sped up or slowed down audio though, that's not great. Besides in the PSX the SPU and CD-ROM are pretty tightly coupled so messing with the timing of one means messing with the timings of the other which can increase the likelihood of game breakage.

Do these games have the same issues when played in "forced NTSC" mode, for instance with PCSX-R? If not that might be a better way to do it, although I would be rather flummoxed if what was the case.

**Comment by Ryunam** [comment]
Okay, I have performed some further testing and comparisons between current Beetle PSX with your commit applied on it and the latest version of the PCSX-ReArmed core.

On Beetle PSX with your PAL 60 patch:
- PAL games run at 60fps;
- The picture has black bars at the top and bottom (this is because, as you previously explained, you avoided altering the timings completely);
- Audio falls out of sync on certain titles, but works perfectly fine on others based on how they were converted by the developers from NTSC to PAL.

On PCSX-ReArmed with Region set to Force NTSC:
- PAL games run at 60fps;
- The picture fills the screen entirely with no black bars;
- I can confirm that the audio falls out of sync with this core as well on those same titles that showed a desync with Beetle PSX. There is a very slight difference in terms of where the audio is breaking, but I assume this is down to the difference in accuracy between the two cores. In general I can say that the desynchronization does happen here as well.

On real PS1 hardware, using a bootdisc that can force video mode (such as Import Player Light):
- PAL games run at 60fps;
- The picture fills the screen entirely with no black bars, however some geometry issues may occur on certain titles (for instance the picture shifts upwards on FFVIII when opening the Menu). This might be a specific bug of the bootdiscs that I have tested;
- The same audio desyncs occur on certain titles, depending on how they were converted.

All of this brings me to say that I would personally like a core option like the one I described above: "Force NTSC timings" with 3 options: OFF (default behaviour), Framerate (your commit), All timings (framerate + no black bars). This would most likely satisfy everyone from a video standpoint.

As for the audio desyncs, unless there is a way to completely prevent them from happening and given that some titles work well without desyncs and some others have this issue, I would find it an acceptable compromise for now.

**Comment by simias** [comment]
Thank you for the extensive testing, that's very useful, at least we know where we stand with this.
It's too bad for the audio sync issues but I strongly suspect that there's no perfect solution that doesn't involve game-specific hacks. Running the video faster is bound to create issues for games that rely on specific video timings to keep the audio in sync. The solution would then be to run the audio itself faster but at this point you might just set your emulator to run ~1.2x faster than normal and get a similar result.

Which renderer did you test with beetle by the way? Only software or hardware as well?

**Comment by Ryunam** [comment]
I have only tested the software renderer on Beetle PSX this time around. I think I remember the OpenGL and Vulkan renderers being a bit more lenient with pre-patched NTSC-forced CD images compared to Beetle PSX software, but I might be mistaken on that.

**Comment by simias** [comment]
I've updated my branch (you'll need to force update if you have the previous version since I rewrote the previous commit): https://github.com/simias/beetle-psx-libretro/tree/dev/pal-60fps

I added an option to override the PAL video timings with 3 values as you suggested:

- No override (same as before, normal PAL timings)

- Fast PAL (PAL resolution, NTSC framerate)

- Force NTSC (we force the GPU to run with full NTSC timings). This one requires a restart of the emulator (fast PAL doesn't, it can be toggled on and off at runtime).

Unfortunately while Fast PAL appears to work correctly, Force NTSC doesn't even manage to boot the BIOS. I must have messed something up or maybe PCSX-R's implementation is not as simple as I surmised.

**Comment by Ryunam** [comment]
Just tested your updated branch and indeed Fast PAL is working perfectly (it's great that it does not require a restart actually), but Force NTSC is not. 
I also tried setting Skip BIOS to ON in the core options, but nothing, the Force NTSC value does not seem to be functioning right now.

**Comment by rz5** [comment]
@Ryunam - Thank you for the real hardware tests. Knowing the benefits and limitations of this hack and how it compares against the same type of hack in other emulators and on a real PS1 will help a lot in bug triaging if and when this gets added.

**Comment by ggdrt** [comment]
> **On PCSX-ReArmed with Region set to Force NTSC:**
> 
> * PAL games run at 60fps;
> * The picture fills the screen entirely with no black bars;
> * I can confirm that the audio falls out of sync with this core as well on those same titles that showed a desync with Beetle PSX. There is a very slight difference in terms of where the audio is breaking, but I assume this is down to the difference in accuracy between the two cores. In general I can say that the desynchronization does happen here as well.

I'm not sure if PCSX ReARMed is a good point of comparison since from what I've seen, the GPU plugins available on it don't seem to properly emulate the display range control commands which is probably why you're not seeing black bars, even for games that are supposed to have them. PCSX is known for audio desynchronization over time in games like the DDR series as it is less timing accurate, which is likely where the differences in audio are coming from.

> **On Beetle PSX with your PAL 60 patch:**
> 
> * PAL games run at 60fps;
> * The picture has black bars at the top and bottom (this is because, as you previously explained, you avoided altering the timings completely);
> * Audio falls out of sync on certain titles, but works perfectly fine on others based on how they were converted by the developers from NTSC to PAL.
> 

Right now the Beetle renderers use the display control values provided by the game itself, so I think that's why you're still seeing black bars rather than it being a timing alteration issue. A soft way to get rid of the black bars would be to use the Initial/Last Scanline core options. Unfortunately the number of scanlines you need to crop for PAL games are going to depend heavily on how the game was converted.

We could probably extend the `Core Aspect Ratio` option to let you force an NTSC-esque aspect ratio while running PAL content so with a little tweaking you can get a somewhat NTSC experience with your PAL content.

**Comment by simias** [comment]
So, if I understood correctly from your tests, our "fast PAL" option and PCSX-R's "force NTSC" option behave fairly similarly in games, the only difference being that "fast PAL" keeps the full PAL resolution (which is bad for games that kept the NTSC resolution with black bars, but good for games that actually bothered to use the additional lines).

If so maybe the simplest solution would be to only have "fast PAL" and then play with the various cropping options to remove the bars when needed?

It would have the advantage of being fairly non-intrusive in the codebase since we already have cropping options (AFAIK, I've never used them) and the fast PAL option only tweaks two lines of codes in the GPU.

Forcing NTSC is a bit messier because it changes many details in the way the GPU functions (different clock divider ratios in particular) which is why in my current non-funcitonal implementation I forced a restart. Savestates can also potentially end up being corrupted if you save in "force_ntsc" mode and load without or vice-versa (it will probably work most of the time, but it's potentially unreliable).

**Comment by simias** [comment]
@ggdrt arg, I should've waited a couple of minutes before replying because you've answered some of my questions:

>I'm not sure if PCSX ReARMed is a good point of comparison since from what I've seen, the GPU plugins available on it don't seem to properly emulate the display range control commands which is probably why you're not seeing black bars, even for games that are supposed to have them.

I should've thought about that, in hindsight it would completely make sense. That means that my force NTSC option in its current implementation is doomed, I'd have to override the display range commands as well.

It may also explain why the games crash: if the display range is greater than the effective resolution the "vsync" IRQ never triggers, that may explain the freeze.

>We could probably extend the Core Aspect Ratio option to let you force an NTSC-esque aspect ratio while running PAL content so with a little tweaking you can get a somewhat NTSC experience with your PAL content.

If it's workable I think it might be the least intrusive and most elegant solution. I'll rework my patch to remove the force NTSC option for now, it's not like it's useful anyway.

**Comment by Ryunam** [comment]
Thank you to both @ggdrt and @simias for the exhaustive answers and for providing more insight about the inner workings of this core and the PS1 hardware itself.

Based on all the information gathered so far, I would agree that the "Fast PAL" option + a new Core Aspect Ratio value for PAL titles with black bars + some title-based tweaking that each user can then apply with the scanline options would basically solve this issue for good. I will be eagerly looking forward to any new development on the matter!

**Comment by simias** [comment]
I've cleaned up my patch, removed the broken "force_ntsc" option and opened a PR.

I don't know if we want to merge that now or wait for the cropping option as well. Technically they're independent from one another (we might want the crop even at normal PAL framerates after all).

**Comment by Papermanzero** [comment]
In terms of feedback I would merge it if it is ready. As you said both features are independent, and users can already try the speed up feature.

**Comment by Ryunam** [comment]
Personally I am definitely in favor of merging it, so that also other people can test it and share their experience. Once we have the implementation of the cropping feature as well, we can then consider this issue officially solved.

**Comment by simias** [comment]
I think that technically I can merge it but I haven't really contributed a lot to this repo in a while so I'd prefer if a more active contributor could review it at least. In particular I've seen that there's been a big refactor for the core option code with internationalization and all that, I don't know if that needs to be updated before merging.

**Comment by ggdrt** [comment]
Fast PAL core option is merged, I'll add the PAL to NTSC aspect ratio conversion soon.

These features might be a bit unstable at first as we get things settled.

**Comment by ggdrt** [comment]
Force NTSC aspect ratio setting under the Core Aspect Ratio option has been added.

I made the line clock counter calculation a bit more accurate using the interlaced rates. The speedup is still a bit off for non-interlaced content though. We might be able to do better by checking the interlace bit and scaling the line clock counter depending on that.

I'm not sure how well this hack works with the way frontends handle audio skew in conjunction with how the core currently reports refresh rate. That could potentially be another source of audio issues.

**Comment by Ryunam** [comment]
Thank you so much for adding both the Fast PAL and NTSC aspect ratio options. I will be sure to now test both features thoroughly and report back in case I notice anything weird or inconsistent.

**Comment by simias** [comment]
Great work @ggdrt, thanks!

**Comment by Ryunam** [comment]
I'm sorry for the really late reply to this! I have been testing the two options and I can confirm the following:
- PAL timing override set to ON works perfectly on all the games I've tested.
- Perhaps I am misinterpreting the Force NTSC aspect ratio option, but I am not sure if it is working as intended: instead of removing the black bars at the top and bottom of the image, this option is currently narrowing the horizontal resolution, making the picture quite thinner compared to the other AR modes available (Corrected, Uncorrected and Force 4:3).
- That being said I've got excellent results by simply tweaking the Initial Scanline and Last Scanline options.

**Comment by ggdrt** [comment]
The progressive PAL signal is ~288 lines at 4:3 aspect ratio while the progressive NTSC signal is ~240 lines at 4:3 aspect ratio. Many games that were ported from NTSC to PAL keep the same internal framebuffer heights and fill in an extra 48 lines of black for PAL conversions, meaning that the PAL ports look squished vertically or stretched horizontally. 

The Force NTSC aspect ratio just unsquishes/unstretches it back to the aspect ratio that you'd see if it were an NTSC game, but doesn't touch the extra PAL lines. Those have to be manually adjusted by the Initial/Last Scanline core options. We don't automatically remove the bars because many NTSC to PAL ports have significant differences in where the image is positioned -- some games don't even vertically center properly. This option shouldn't be needed for games that actually take advantage of PAL height like Spyro or Crash Bandicoot.

**Comment by simias** [comment]
It's getting off-topic but mednafen's emulation of proper TV timings seems more trouble than it's worth in my experience. Very few games actually benefit from emulating that properly (as far as I know mainly "screen shake" effects in a few games). It's probably good to have for those games but it makes for a lousy default. We basically have code that bothers to add black lines and then we add options to manually get rid of them. That's a bit silly.

**Comment by Ryunam** [comment]
Thanks for the explanation, @ggdrt. I was under the (wrong) impression that the Force NTSC option was supposed to eliminate the top-bottom black bars, but it makes perfect sense how that can't be an automatic thing and I now understand what purpose is served by the aspect ratio feature.

I will keep this issue open for a little while longer, in order to let others report any observation / finding or lingering issues, but I think that with a combination of
- PAL timing override for 60Hz gameplay,
- Force NTSC aspect ratio, and
- Scanline options to tweak and remove the black bars
PAL games can now be enjoyed to their fullest.

Thank you all for the amazing contributions so far and especially for the hard work put towards fulfilling this long-standing request.

**Comment by BardBun** [comment]
I'm glad this option is finally available, however it doesn't work as one would wish for it to work for PAL games:
https://www.youtube.com/watch?v=COZMT2TzJ5M

PAL 2 NTSC patched games on other emulators will work and behave exactly like the NTSC part in the video.
All this Override option does is speed up the game for no reason.
The "slower PAL" thing people always say seems to come from the SNES era, however on more "modern" consoles what happens is that a PAL game leaves 5 frames (25 instead of 30) out, ingame this leads to virtually no difference however in scripted (fps dependent) cutscenes as well as FMVs this will lead to a sort of "teleporting" visual, an example can be seen here: https://www.youtube.com/watch?v=HPuBjkFYoOw&t=171s

If you can't see it that easily, simply lower the videos speed to 0.25s and you will see that every second the video skips a few frames, leading to a cut like / teleport effect caused by it being the PAL version.

A PAL 2 NTSC patched game on an emulator that accepts it will not have any of those issues and will behave just like the original NTSC version.

Simply speeding up the PAL version to 30 FPS does not get rid of the overall issues and will only cause games to either run too fast or cutscenes to break.

**Comment by trostboot** [comment]
> The "slower PAL" thing people always say seems to come from the SNES era, however on more "modern" consoles what happens is that a PAL game leaves 5 frames (25 instead of 30) out, ingame this leads to virtually no difference

Citation needed. There are many, many, _many_ 5th and 6th generation console games that are simply slowed down and squished for the PAL version, because game speed being tied to FPS was commonplace.
Proper adaptations are a rarity. I remember Gran Turismo 4 being praised for being a rare example of a decent port, and that's a mid-to-late 6th gen game.

**Comment by Ryunam** [comment]
Closing this for now, as the feature has been implemented in this core for quite some time and so far has been working reliably, albeit with a few inconsistencies that inevitably derive from the mere fact of forcing a game hard-coded at 50hz to play at 60hz (ie de-synced movies in certain instances, or the skips that were mentioned before). These issues also happen on real hardware whenever using any software / serial-port device that forces a specific video format and they're also highly game-dependent. 

I speak from direct experience and a lot of testing when I say that *lot* of PAL PS1 games were not optimized at all, so having this option is highly beneficial in all such cases and the issues described above do not constitute an example of this feature possibly malfunctioning. 

Thank you to everyone involved for adding this override!
